### PR TITLE
Implementing is_compatible

### PR DESCRIPTION
**Pull request by petr-tik**
Opening this now to get feedback -  please don't merge yet. 

I have only found Footer::extract_footer used outside the module, so decided to hide Footer::from_bytes from public consumption

Add a new error type for IncompatibleIndex
add a prototypical call to footer.is_compatible() in ManagedDirectory::open_read to make sure we error before reading it further

### CODE REVIEW

**Comment by petr-tik** [comment]
2 questions:

1. Is this compatibility logic correct? 
2. The [stacktrace](https://github.com/tantivy-search/tantivy/issues/692#issue-520006546) from #692 shows ManagedDirectory::open_read, which we currently deal with.

However, the local lib_search library [uses](https://gitlab.com/crates.rs/crates.rs/blob/master/search_index/src/lib_search_index.rs#L49) Index::open_in_dir, which calls into MmapDirectory, which currently doesn't create or check footers.

Therefore, this fix won't help people to check the compatibility of indexes they open through Index::open_in_dir. 

I don't think it's a great idea to surface the footer interface all the way to the user and i welcome any advice/recommendation on where to fit Footer::is_compatible in MmapDirectory::open_read.

**Comment by petr-tik** [comment]
ManagedDirectory::wrap uses Directory::atomic_read, which looks like a better place to do footer compatibility checking (or is it both?).

File: src/directory/footer.rs [hunk 1]
```diff
@@ -102,6 +105,27 @@ impl Footer {
     pub fn size(&self) -> usize {
         self.versioned_footer.size() as usize + self.meta.len() + COMMON_FOOTER_SIZE
     }
+
+    /// Compares the version found in the footer with the version of tantivy running
+    /// Indexes are incompatible when the major and minor versions are different between the
+    /// current library and serialised index are different
```
Comment (fulmicoton) [inline src/directory/footer.rs]: This is incorrect. We now have two versions in the Footer. 

Tantivy's version is here for debugging purpose. What is important is the index version. Right now only 0 is available.

File: src/lib.rs [hunk 2]
```diff
@@ -171,8 +171,9 @@ pub use crate::postings::Postings;
 pub use crate::reader::LeasedItem;
 pub use crate::schema::{Document, Term};
 
-/// Expose the current version of tantivy, as well
-/// whether it was compiled with the simd compression.
+/// Expose the current version of tantivy as found in Cargo.toml during compilation.
+/// eg. "0.11.0"
+/// as well whether it was compiled with the simd compression.
```
Comment (petr-tik) [inline src/lib.rs:176]: @fulmicoton Does this method give any meaningful info about simd compression in tantivy? Does this refer to snappy vs lz4 compression methods? 

If it doesn't - should it?

Comment (fulmicoton) [reply src/lib.rs:176]: Ah no the SIMD copmression bit is irrelevant today. There used to be a different codec, but now tantivy uses a scalar fixture when the instruction set is not available. All platform use the same codec.

**Pull request closed without merging**